egs/vivos/asr1/RESULTS.md

```diff
@@ -1,120 +1,487 @@
-# pytorch Transducer (enc: VGG2L + 6 x Transformer, dec: 1 x LSTM)
+# RNN-Transducer (enc: 4 x BLSTMP, dec: 1 x LSTM)
 
 - Environments
-  - date: `Fri Apr  3 10:34:40 CEST 2020`
+  - date: `Sat Aug 22 16:14:35 CEST 2020`
   - python version: `3.7.3 (default, Mar 27 2019, 22:11:17)  [GCC 7.3.0]`
   - espnet version: `espnet 0.6.2`
   - chainer version: `chainer 6.0.0`
   - pytorch version: `pytorch 1.0.1.post2`
-  - Git hash: `d19105b0793b1a9c6e11293f2f95f85699632367`
-  - Commit date: `Fri Apr 3 10:09:03 2020 +0200`
-
-- Model files (archived to model.tar.gz by `$ pack_model.sh`)
-  - model link: https://drive.google.com/file/d/1mW2ANumwGozam7mt2_0JSpUAWIFfvyac
-  - training config file: `conf/tuning/transducer/train_transformer-rnn_transducer.yaml`
-  - decoding config file: `conf/tuning/transducer/decode_transducer.yaml`
-  - cmvn file: `data/train_nodev/cmvn.ark`
-  - e2e file: `exp/train_nodev_pytorch_train_transformer-rnn_transducer/results/model.last5.avg.best`
-  - e2e JSON file: `exp/train_nodev_pytorch_train_transformer-rnn_transducer/results/model.json`
-  - lm file: `exp/train_rnnlm_pytorch_lm_word7184/rnnlm.model.best`
-  - lm JSON file: `exp/train_rnnlm_pytorch_lm/model.json`
-  - dict file: `data/lang_1char`
-
-- Results (obtained with `$ show_results.sh`)
+  - Git hash: `077c8970afb477e059932f7243a9b728c8ab1a69`
+  - Commit date: `Fri Aug 21 16:32:27 2020 +0200`
 
 ## CER
 
 |dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
 |---|---|---|---|---|---|---|---|---|
-|decode_test_decode_transducer|760|32771|85.4|10.8|3.8|3.0|17.6|94.2|
-|decode_test_decode_transducer_lm|760|32771|85.6|10.6|3.8|3.0|17.4|93.3|
-|decode_test_decode_transducer_lm_word7184|760|32771|86.1|10.2|3.7|2.8|16.7|92.1|
-|decode_train_dev_decode_transducer|100|4007|84.7|12.8|2.6|3.2|18.6|97.0|
-|decode_train_dev_decode_transducer_lm|100|4007|84.7|12.6|2.7|3.3|18.6|97.0|
-|decode_train_dev_decode_transducer_lm_word7184|100|4007|84.9|12.4|2.7|3.3|18.4|97.0|
+|decode_test_decode_alsd|760|32771|85.3|11.5|3.2|3.6|18.3|93.8|
+|decode_test_decode_alsd_lm|760|32771|85.7|11.1|3.2|3.6|17.9|93.0|
+|decode_test_decode_alsd_lm_word7184|760|32771|86.2|10.7|3.1|3.5|17.3|92.0|
+|decode_test_decode_default|760|32771|84.6|11.6|3.8|3.2|18.6|93.3|
+|decode_test_decode_default_lm|760|32771|84.1|10.9|5.0|2.9|18.8|93.0|
+|decode_test_decode_default_lm_word7184|760|32771|84.4|10.7|4.9|2.8|18.4|92.0|
+|decode_test_decode_nsc|760|32771|85.2|11.4|3.4|3.5|18.3|93.0|
+|decode_test_decode_nsc_lm|760|32771|85.5|11.1|3.4|3.4|17.9|92.5|
+|decode_test_decode_nsc_lm_word7184|760|32771|85.9|10.8|3.3|3.3|17.4|92.0|
+|decode_test_decode_tsd|760|32771|85.0|11.4|3.5|3.2|18.2|93.6|
```

Comment:
Would you note the decoding parameter somewhere? e.g. The n of `nsc`.

Reply:
The meaning of "N" is explained in the tutorial part alongside other algorithm parameters and the `n-step` parameter is given in `decode_nsc.yaml` config, is that okay?

Reply:
Did you test n==3?
I didn't read the papers detail, but just only according to your report, I understood that only `NSC [nstep=1, prefix-alpha=1]` or `ALSD [u-max=20]` are faster than your new default, so I'd like to the result, too.

Reply:
Yes, I tested N=[1, 3] for NSC and max_sym_exp=[2, 4] for TSD (note: N=(max_sym_exp+1)) . The results are for respectively N=3 and max_sym_exp=4.

I'll re-run experiments for N=1 and max_sym_exp=2 but from what I remember CER is about 20% for NSC and 60% for TSD with vanilla RNNT. The difference between the two can be explained by a condition in NSC I added, related to blank.

Reply:
It should be noted that with most corpora (such as Librispeech, WSJ, ESTER, etc) you would use N=1 and max_sym_exp=2. Vivos is a one of the special "case" which motivated the modification of OSC into NSC

Reply:
With vanilla RNN-T (performance difference for NSC and TSD with same parameters would be the same across models):

| Algorithm [beam=10] | CER (test) | WER (test) |
| - | - | - |
| Default (new) | 18.6 | 39.0 |
| NSC [nstep=1, prefix-alpha=1] | 24.3 | 50.0 |
| NSC [nstep=3, prefix-alpha=2] | 18.3 | 38.7 |
| ALSD [u-max=10] | 33.2 | 50.8 | 
| ALSD [u-max=100] | 18.3 | 38.4 |
| TSD [max_sym_exp=2] | 64.4 | 80.1 |
| TSD [max_sym_exp=4] | 18.2 | 38.5 |

I also forgot to mention other things:
- The condition I added in NSC is for N=1 is not complete in the espnet version: it shouldn't be applied depending on other parameters (e.g.: if monoRNNT is used, if a prior validation step give a theoretical N equal to 1, etc). However, I confirmed on several dataset (Librispeech, WSJ, french corpus, etc) that even in this cases the condition doesn't decrease performance in any ways (and could even increase).
- If I add the condition to TSD algorithm, I can reach similar performance for max_sym_exp=2 vs N=1.